design-documents/web-api-controllers.md

```diff
@@ -0,0 +1,320 @@
+# Middleware for web APIs, user input processing on top of service contracts execution
+## RESTful and SOAP API
+#### Current situation
+Service contracts, their arguments and return types are supposed to be used directly
+as web API endpoints. The problem is service contracts are also used as API
+and are supposed to be a part of business layer. By using them for presentation developers may encounter various
+problems like having to put additional logic related to specifically processing web API requests (fields validation
+for instance) inside models, DTOs used as arguments and return types may have sensitive data that is fine to use programmatically but
+should not be exposed to web APIs or simply having properties needed for implicit business logic that would confuse web
+API consumers.
+ 
+###### Example of service contracts not being tailored to be used directly as web API endpoints
+For instance let's take a look at creating a customer account:
+ 
+```php
+    AccountManagementInterface::createAccount(
+        \Magento\Customer\Api\Data\CustomerInterface $customer,
+        $password = null,
+        $redirectUrl = ''
+    );
+```
+ 
+The service contract responsible for it has this method signature, right off the bat we can see an argument that should
+not be accessible to web API - _redirectUrl_ which is used to set redirect URL inside confirmation E-mail.
+And what about password? From a web API client's perspective it should be the $customer's property but it's a separate
+account because _CustomerInterface_ is meant to represent business layer entity which doesn't store original password
+and only has the password's hash. Also _CustomerInterface_ has _defaultShipping/Billing_ properties, but then
+_AddressInterface_ also has _isDefaultShipping/Billing_ properties - which one should a web API client use to set a
+customer's default billing/shipping address? Another problem - customer's addresses have _customerId_ property which
+doesn't make sense in context of a customer managing itself since a customer can manage only their own addresses.
+All of these problems stem from having business layer contracts/entities exposed as presentation layer operations/data.
+ 
+ 
+###### Examples of classes that contain presentation layer logic
+While service contracts implementations usually can contain all the logic needed to execute an operation various
+presentation areas would require additional logic to convey results to users or to extract operations' arguments from
+user input. Since service contracts are not supposed to know about context they are also not feet to contain additional
+authorization related logic.
+ 
+Examples of cases when we need additional context aware/presentation logic:
```

Comment:
These are not use cases, but assumptions that some use cases need this logic.
Can you please provide use cases (scenarios)? For example, "When updating customer address from SF, we should validate that the customer owns it". For example, this use case should not need any additional framework implementation. W/o real use cases we might be adding features that will move us in the direction opposite to the desired state.

So far, classes that are needed for presentation (e.g., data filtering/conversion) should be just located somewhere close to that presentation (controller, block, etc). And other cases (e.g., ownership validation) can either be solved differently or should not be considered basis for generalization (it's just a feature that should be implemented as is on the framework level).